src/sys/windows/named_pipe.rs

```diff
@@ -0,0 +1,610 @@
+//! implementation on Windows.
+//!
+//! # Trait implementations
+//!
+//! The `Read` and `Write` traits are implemented for `NamedPipe` and for
+//! `&NamedPipe`. This represents that a named pipe can be concurrently read and
+//! written to and also can be read and written to at all. Typically a named
+//! pipe needs to be connected to a client before it can be read or written,
+//! however.
+//!
+//! Note that for I/O operations on a named pipe to succeed then the named pipe
+//! needs to be associated with an event loop. Until this happens all I/O
+//! operations will return a "would block" error.
+//!
+//! # Managing connections
+//!
+//! The `NamedPipe` type supports a `connect` method to connect to a client and
+//! a `disconnect` method to disconnect from that client. These two methods only
+//! work once a named pipe is associated with an event loop.
+//!
+//! The `connect` method will succeed asynchronously and a completion can be
+//! detected once the object receives a writable notification.
+//!
+//! # Named pipe clients
+//!
+//! Currently to create a client of a named pipe server then you can use the
+//! `OpenOptions` type in the standard library to create a `File` that connects
+//! to a named pipe. Afterwards you can use the `into_raw_handle` method coupled
+//! with the `NamedPipe::from_raw_handle` method to convert that to a named pipe
+//! that can operate asynchronously. Don't forget to pass the
+//! `FILE_FLAG_OVERLAPPED` flag when opening the `File`.
+
+use crate::{poll, Registry};
+use crate::sys::windows::Overlapped;
+
+use std::ffi::OsStr;
+use std::fmt;
+use std::io::{self, Read};
+use std::mem;
+use std::os::windows::io::*;
+use std::slice;
+use std::sync::atomic::{AtomicUsize, AtomicBool};
+use std::sync::atomic::Ordering::{Relaxed, SeqCst};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+// use mio::windows;
+// use mio::{Evented, Poll, PollOpt, Ready, Registration, SetReadiness, Token};
+use miow::iocp::CompletionStatus;
+use miow::pipe;
+use winapi::shared::winerror::*;
+use winapi::um::ioapiset::*;
+use winapi::um::minwinbase::*;
+
+macro_rules! offset_of {
+    ($t:ty, $($field:ident).+) => (
+        &(*(0 as *const $t)).$($field).+ as *const _ as usize
+    )
+}
+
+macro_rules! overlapped2arc {
+    ($e:expr, $t:ty, $($field:ident).+) => ({
+        let offset = offset_of!($t, $($field).+);
+        debug_assert!(offset < mem::size_of::<$t>());
+        Arc::from_raw(($e as usize - offset) as *mut $t)
+    })
+}
+
+fn would_block() -> io::Error {
+    io::ErrorKind::WouldBlock.into()
+}
+
+/// Representation of a named pipe on Windows.
+///
+/// This structure internally contains a `HANDLE` which represents the named
+/// pipe, and also maintains state associated with the mio event loop and active
+/// I/O operations that have been scheduled to translate IOCP to a readiness
+/// model.
+pub struct NamedPipe {
+    inner: Arc<Inner>,
+}
+
+struct Inner {
```

Comment:
I believe for this to work with `overlapped2arc`, or more specifically `offset_of`, this should be `#[repr(C)]`.

See https://github.com/Gilnaa/memoffset/pull/43, https://github.com/rust-lang/rust/issues/48956#issuecomment-544506419 and https://github.com/rust-lang/rust/issues/64490#issuecomment-657089537.